README.md
![image](https://user-images.githubusercontent.com/2525601/142946020-7ac32b9d-3fb8-4607-9c20-169623fe4513.png)

# Overview
Simple transaction tracker website.
No fluff, just simple transaction info.

I really hate most of the crypto portfolio apps out there, both paid and free.
They don't handle transactions cleanly and who knows what they are doing behind the scenes.

The goal of this project is to provide an open source web based portfolio manager that anyone can use.

# Tech Stack
- Django
- Django Rest Framework
- Python
- Javascript
- HTML5
- JQuery
- Bootstrap
- SQLLite
- AdminLTE 3 (aka the best html5 theme out there)

# Features
- All data endpoints are available on `http://localhost:8000/api`
- Self contained (no external dependencies besides coinbase api call for price lookups)
- No API keys required (for now)
- Transaction Importers for
  - Blockfi
  - Coinbase
  - Coinbase Pro
  - Celcius
  - Crypto.com
- Simple data tables with full export
- Auth on by default (google auth can be turned on, local django auth works fine too)
- Forever free, always open source

# API
- Browsable (mostly) API
  - http://localhost:8000/api
- Realtime and historical price lookup:
  - http://localhost:8000/api/spot-price/btc/2021-04-30-15-15-30/
- Transactions
  - http://localhost:8000/api/transactions

# How to support
- PR requests always welcome
- Dontate via BTC!
```
BTC Address: 3DDqESKCb6nXeNbEniqsmCSgzHd2Sk6nuU
```

# Requirements
- Python 3.9+

# Getting Started
- For first time run: `./initialize.sh` (only needed to run once)
- Django web server can be started any time after that via vscode debugger `launch.json` included or manually via `python ./main/manage.py runserver`

# Changelog
[Current](CHANGELOG.md)


CHANGELOG.md
# Changelog

## 0.22.0 (Dec 30, 2021)

FEATURES:
* Upgraded support to python 3.10
* Refactor importers
* Add `pytest` and tests finally!
* home dashboard with realtime prices and aggregate data!

ENHANCEMENTS:

BUG FIXES:

---

## 0.21.0 (Dec 28, 2021)

FEATURES:
* Added asset summary API call (ex: `http://localhost:8000/api/asset-info/BTC/`)
* Added `needs_reviewed` to `transactions` to enable easier review statusing

ENHANCEMENTS:

BUG FIXES:

---

## 0.20.0 (Nov 29, 2021)

FEATURES:
* Added ability to modify cryptkeeper.csv export and merge/match it back on import. This allows modifying locally for bulk operations.
* `import_hash` exported from datatable now to support this

ENHANCEMENTS:

BUG FIXES:

KNOWN ISSUES:
* Saving csv in excel wipes seconds from timestamp which causes `updates` to occur and loss of precision. Not a cryptkeeper problem but will probably cause some headaches.

---

## 0.19.0 (Nov 29, 2021)

FEATURES:
* Added parsers for `crypto.com`, `coinbase pro`, and `celcius`

ENHANCEMENTS:

BUG FIXES:
* Fixed `recieve` total quantity on model

---

## 0.18.0 (Nov 28, 2021)

FEATURES:

ENHANCEMENTS:
* Moved parsers to core, renamed, and de-coupled from django dependencies

BUG FIXES:

---

## 0.17.1 (Nov 24, 2021)

FEATURES:

ENHANCEMENTS:
* Converted coinbase importer to be more standard

BUG FIXES:
* Fixed more rounding oddness
* Fixed `./initialize.sh` to actually work

---

## 0.17.0 (Nov 24, 2021)

FEATURES:
* Added CryptKeeper csv importer
  * This is akin to a backup, export the CSV from transactions and you can import it back in on the importer page

ENHANCEMENTS:
* Fixed some decimal logic and moved validators a bit
* TODO: DFR seems to 

BUG FIXES:

Known Issues:
* DFR won't trim trailing zeros from the API and it causes so strangeness. Had to increase zeros for now.

---


## 0.16.1 (Nov 23, 2021)

FEATURES:

ENHANCEMENTS:

BUG FIXES:
* /api rooted incorrectly

---

## 0.16.0 (Nov 23, 2021)

FEATURES:
* Spot price API frontend created
* Added support for more time formats to DRF

ENHANCEMENTS:

BUG FIXES:
---

## 0.15.1 (Nov 22, 2021)

FEATURES:

ENHANCEMENTS:
* Move changelog to dedicated file
* Updated welcome page.

BUG FIXES:
---

## 0.15.0 (Nov 22, 2021)

FEATURES:
* `crypto_price_finder` added for historical price lookup, no api key required
  * Note: It is running off of coinbase's public API with throttling (around 10/second). Should be fine for light workloads
* Blockfi `blockfi_all_transactions` importer created. Should fully cover all blockfi transactions if you upload both reports

ENHANCEMENTS:
* Added better auto-detection for reports
* Expanded `notes` field to 1000 characters

BUG FIXES:
* Bad included for `dotmap` which is no longer used

---

## 0.14.0 (Nov 14, 2021)

FEATURES:
* Rename transaction fields

---

## 0.13.0 (Nov 14, 2021)

FEATURES:
* Refactor coinbase importer to support convert
* Added type `send` and `airdrop`
* Added export all for transcations


ENHANCEMENTS:

BUG FIXES:

---

## 0.12.0 (Oct 27, 2021)

FEATURES:
* Blockfi importer partially working
* Added `import_hash` field to `transaction` to allow users to edit fields and re-importing doesn't override them
* Added `delete all transcations` capability

ENHANCEMENTS:

BUG FIXES:

---

## 0.11.0 (Oct 21, 2021)

FEATURES:
* API uploader front end converted to ajax
* File importer returns row counts
* Fixed icons
* Mockup of charting on home page

ENHANCEMENTS:

BUG FIXES:

---

## 0.10.0 (Oct 20, 2021)

FEATURES:
* Coinbase importer mockup working
* `Fee` field added
* Multi-select delete working on `transactions`

ENHANCEMENTS:

BUG FIXES:

---

## 0.9.0 (Oct 19, 2021)

FEATURES:
* Multi select exclusion of last row to avoid buttons causing selection
* Working upload API

ENHANCEMENTS:

BUG FIXES:
* Removed .ds_store files

---

## 0.8.0 (Oct 19, 2021)

FEATURES:
* Multi select on datatables working
* Fixed some styling for datatables

ENHANCEMENTS:

BUG FIXES:

---

## 0.7.0 (Oct 18, 2021)

FEATURES:
* `PUT`, `UPDATE`, `DELETE` all implemented
* UI CRUD for transactions working
* TODO: Cleanup JS and write tests

ENHANCEMENTS:

BUG FIXES:

---

## 0.6.0 (Oct 16, 2021)

FEATURES:
* Working `put` example with toast
* TODO: lock down the rest of the api methods and cleanup UI

ENHANCEMENTS:

BUG FIXES:

---

## 0.5.0 (Oct 14, 2021)

FEATURES:
* Datatables rest api with django wired up and working transactions page
* TODO: Cleanup buttons and ajax crud from js
* Added `from` an `to` for into model `transaction`

ENHANCEMENTS:

BUG FIXES:
* Admin registration fix

---

## 0.4.0 (Oct 12, 2021)

FEATURES:
* Django Rest Framework wired up. Will need cleaned up arrangement
* Basic user data filtered rest api added

ENHANCEMENTS:

BUG FIXES:

---

## 0.3.0 (Oct 12, 2021)

FEATURES:
* Working user data association
* Base template created.

ENHANCEMENTS:

BUG FIXES:

---

## 0.2.0 (Oct 7, 2021)

FEATURES:
* `Adminlte3` theme loaded!
* Structure adjusted a bit to be best practice

ENHANCEMENTS:

BUG FIXES:
* `.gitignore` updates

---

## 0.1.1 (Oct 7, 2021)

FEATURES:

ENHANCEMENTS:
* Added `initialize.sh` to generate the secret key
* Added to getting started

BUG FIXES:
* removed secret key from `main/settings.py` and replaced with `secret_key.txt` include (which is in .gitingore)
---

## 0.1.0 (Oct 7, 2021)

FEATURES:
* README.md setup
* Initial cut
* Google SSO working
* Basic transaction model setup

ENHANCEMENTS:

BUG FIXES:

main/cryptkeeper/api.py
from django.db.models.aggregates import Count
from django.db.models.fields import FloatField
from rest_framework import serializers, viewsets, permissions
from rest_framework.response import Response
from rest_framework.decorators import action
import json
from django.db.models import Sum, Q, F
from django.db.models.functions import Coalesce, Abs
from django.core import serializers as django_serializer
from django.http import JsonResponse
from .core.cryptkeeper import tools as cryptkeeper_tools
from .core.crypto_price_finder import crypto_price_finder
from .models import *


#
# Transactions
# 
class TransactionSerializer(serializers.ModelSerializer):
    class Meta:
        model = Transaction
        exclude = ['user']

    def create(self,validated_data):
        request = self.context['request']
        transaction = Transaction.objects.create(**validated_data, user = request.user)
        transaction.save()
        return transaction

class TransactionViewSet(viewsets.ModelViewSet):
    serializer_class = TransactionSerializer
    queryset = Transaction.objects.all()
    permission_classes = [permissions.IsAuthenticated]

    def get_queryset(self):
        return Transaction.objects.filter(user=self.request.user)

#
# File Importer
#
class TransactionImporterSerializer(serializers.Serializer):
    file = serializers.FileField()

    class Meta:
        fields = ('file')

class TransactionImporterViewSet(viewsets.ViewSet):
    serializer_class = TransactionImporterSerializer
    permission_classes = [permissions.IsAuthenticated]  

    def create(self, request, *args, **kwargs):
        file        = request.FILES.get('file')
        file_name   = file.name

        results = cryptkeeper_tools.import_transactions_from_file(
            file_name       = file_name, 
            in_memory_file  = file, 
            user            = self.request.user
        )

        return Response(json.dumps(results))

#
# Spot Price
#
class SpotPriceSerializer(serializers.Serializer):
    asset_symbol = serializers.CharField()
    datetime     = serializers.DateTimeField(format="%Y-%m-%d-%H:%M:%S")

class SpotPriceViewSet(viewsets.ViewSet):
    serializer_class = SpotPriceSerializer
    permission_classes = [permissions.IsAuthenticated]

    def list(self, request):
        return Response([])

    #ex: http://localhost:8000/api/spot-price/btc/2020-04-30-15-15-30/
    @action(methods=['get'], detail=True, url_path='(?P<datetime>[^/.]+)', serializer_class= SpotPriceSerializer())
    def get_spot_price(self, request, datetime, pk=None):
        serializer = SpotPriceSerializer(data={'asset_symbol': pk, 'datetime': datetime})
        serializer.is_valid(raise_exception=True)

        price_info = crypto_price_finder.get_usd_price(
            target_time     = serializer.validated_data["datetime"],
            asset_symbol    = serializer.validated_data["asset_symbol"]
        )

        if not price_info["success"]:
            return Response("Request could not be completed.", status=400)
        
        return Response(price_info)

#
# Asset Info
# 
class AssetInfoViewSet(viewsets.ViewSet):
    permission_classes = [permissions.IsAuthenticated]

    def list(self, request):
        asset_transactions = Transaction.objects                                                                                     \
            .filter(user=request.user)                                                                                               \
            .values("asset_symbol")                                                                                                  \
            .annotate(total_asset_quantity             = Sum("asset_quantity"))                                                      \
            .annotate(total_asset_quantity_buy         = Sum("asset_quantity"           , filter=Q(transaction_type="Buy")))         \
            .annotate(total_asset_quantity_sell        = Sum("asset_quantity"           , filter=Q(transaction_type="Sell")))        \
            .annotate(total_asset_quantity_interest    = Sum("asset_quantity"           , filter=Q(transaction_type="Interest")))    \
            .annotate(total_asset_quantity_airdrop     = Sum("asset_quantity"           , filter=Q(transaction_type="Airdrop")))     \
            .annotate(total_usd_buy                    = Sum("usd_total_with_fees"      , filter=Q(transaction_type="Buy")))         \
            .annotate(total_usd_sell                   = Sum("usd_total_with_fees"      , filter=Q(transaction_type="Sell")))        \
            .annotate(total_transacted_usd             = Sum("usd_total_with_fees"))                                                 \
            .annotate(total_transactions               = Count("usd_total_with_fees"))                                               \
            .annotate(average_price_buy                = Abs(F("total_usd_buy") / F("total_asset_quantity_buy")))

        data = list(asset_transactions)
        return Response(data)

    #ex: http://localhost:8000/api/asset-info/btc/
    def retrieve(self, request, pk=None):
        return Response([])